cmd/launcher/extension.go

```diff
@@ -0,0 +1,156 @@
+		}
+		enrollSecret = string(bytes.TrimSpace(content))
+	}
+
+	// create the certificate pool
+	var rootPool *x509.CertPool
+	if opts.rootPEM != "" {
+		rootPool = x509.NewCertPool()
+		pemContents, err := ioutil.ReadFile(opts.rootPEM)
+		if err != nil {
+			return nil, errors.Wrapf(err, "reading root certs PEM at path: %s", opts.rootPEM)
+		}
+		if ok := rootPool.AppendCertsFromPEM(pemContents); !ok {
+			return nil, errors.Errorf("found no valid certs in PEM at path: %s", opts.rootPEM)
+		}
+	}
+
+	// connect to the grpc server
+	grpcConn, err := service.DialGRPC(opts.kolideServerURL, opts.insecureTLS, opts.insecureGRPC, opts.certPins, rootPool, logger)
+	if err != nil {
+		return nil, errors.Wrap(err, "dialing grpc server")
+	}
+
+	// create the client of the grpc service
+	launcherClient := service.New(grpcConn, level.Debug(logger))
+
+	// create the osquery extension
+	extOpts := osquery.ExtensionOpts{
+		EnrollSecret:    enrollSecret,
+		Logger:          logger,
+		LoggingInterval: opts.loggingInterval,
+	}
+
+	// create the extension
+	ext, err := osquery.NewExtension(launcherClient, db, extOpts)
+	if err != nil {
+		return nil, errors.Wrap(err, "starting grpc extension")
+	}
+
+	// create the logging adapter for osquery
+	osqueryLogger := &kolidelog.OsqueryLogAdapter{Logger: level.Debug(log.With(logger, "component", "osquery"))}
+
+	var runner *runtime.Runner
+
+	return &extension{
+		// add the runtime to the extension
+		Runner: runner,
+		// and the methods for starting and stopping the extension
+		Execute: func() error {
+			println("\nextension started\n")
+			// Start the osqueryd instance
+			runner, err = runtime.LaunchInstance(
+				runtime.WithOsquerydBinary(opts.osquerydPath),
+				runtime.WithRootDirectory(rootDirectory),
+				runtime.WithConfigPluginFlag("kolide_grpc"),
+				runtime.WithLoggerPluginFlag("kolide_grpc"),
+				runtime.WithDistributedPluginFlag("kolide_grpc"),
+				runtime.WithOsqueryExtensionPlugin(config.NewPlugin("kolide_grpc", ext.GenerateConfigs)),
+				runtime.WithOsqueryExtensionPlugin(osquerylogger.NewPlugin("kolide_grpc", ext.LogString)),
+				runtime.WithOsqueryExtensionPlugin(distributed.NewPlugin("kolide_grpc", ext.GetQueries, ext.WriteResults)),
+				runtime.WithOsqueryExtensionPlugin(table.LauncherIdentifierTable(db)),
+				runtime.WithStdout(osqueryLogger),
+				runtime.WithStderr(osqueryLogger),
+				runtime.WithLogger(logger),
+			)
+			if err != nil {
+				logger.Fatal(errors.Wrap(err, "launching osquery instance"))
+			}
+
+			// The runner allows querying the osqueryd instance from the extension.
+			// Used by the Enroll method below to get initial enrollment details.
+			ext.SetQuerier(runner)
+
+			_, invalid, err := ext.Enroll(ctx)
+			if err != nil {
+				logger.Fatal("err", errors.Wrap(err, "enrolling host"))
+			}
+			if invalid {
+				logger.Fatal(errors.Wrap(err, "invalid enroll secret"))
+			}
+			ext.Start()
+
+			// TODO: remove when underlying libs are refactors
```

Comment:
:+1: We could also make `ext.Start` take a context and have it block on this channel if you want `Start()` to block.